tests/struct_test.py

```diff
@@ -0,0 +1,235 @@
+    expr = df.array.struct.get("col1")
+    assert expr.tolist() == [1, 2, 3]
+
+    expr = df.array.struct.get("col2")
+    assert expr.tolist() == ["a", "b", "c"]
+
+
+def test_struct_get_label_duplicated_raise(df_duplicated):
+    with pytest.raises(ValueError):
+        df_duplicated.array.struct.get("col1").tolist()
+
+
+def test_struct_get_index(df):
+    expr = df.array.struct.get(0)
+    assert expr.tolist() == [1, 2, 3]
+
+    expr = df.array.struct.get(1)
+    assert expr.tolist() == ["a", "b", "c"]
+
+
+def test_struct_get_index_duplicated(df_duplicated):
+    expr = df_duplicated.array.struct.get(0)
+    assert expr.tolist() == [1, 2, 3]
+
+    expr = df_duplicated.array.struct.get(1)
+    assert expr.tolist() == ["a", "b", "c"]
+
+
+def test_struct_get_getitem_notation(df):
+    assert df.array[:, "col1"].tolist() == [1, 2, 3]
+    assert df.array[:, "col2"].tolist() == ["a", "b", "c"]
+    assert df.array[:, 0].tolist() == [1, 2, 3]
+    assert df.array[:, 1].tolist() == ["a", "b", "c"]
+
+
+def test_struct_get_invalid_field(df):
+    with pytest.raises(ValueError):
+        df.array.struct.get("doesNotExist").tolist()
+
+
+def test_struct_get_invalid_dtype(df):
+    """Ensure that struct function is only applied to correct dtype."""
+    with pytest.raises(TypeError):
+        df.integer.struct.get("col1").tolist()
+
+    with pytest.raises(TypeError):
+        df.integer.struct.get(0).tolist()
+
+
+def test_struct_project_label(df):
+    expr = df.array.struct.project(["col3", "col1"])
+    assert expr.tolist() == [{"col3": 4, "col1": 1},
+                             {"col3": 5, "col1": 2},
+                             {"col3": 6, "col1": 3}]
+
+
+def test_struct_project_index(df):
+    expr = df.array.struct.project([2, 0])
+    assert expr.tolist() == [{"col3": 4, "col1": 1},
+                             {"col3": 5, "col1": 2},
+                             {"col3": 6, "col1": 3}]
+
+
+def test_struct_project_mixed_label_index(df):
+    expr = df.array.struct.project(["col3", 0])
+    assert expr.tolist() == [{"col3": 4, "col1": 1},
+                             {"col3": 5, "col1": 2},
+                             {"col3": 6, "col1": 3}]
+
+
+def test_struct_project_getitem_notation(df):
+    assert df.array[:, ["col3", "col1"]].tolist() == [{"col3": 4, "col1": 1},
+                                                      {"col3": 5, "col1": 2},
+                                                      {"col3": 6, "col1": 3}]
+
+    assert df.array[:, [2, 0]].tolist() == [{"col3": 4, "col1": 1},
+                                            {"col3": 5, "col1": 2},
+                                            {"col3": 6, "col1": 3}]
+
+
+def test_struct_project_invalid_field(df):
+    with pytest.raises(ValueError):
+        df.array.struct.project(["doesNotExist"]).tolist()
+
+
+def test_struct_project_invalid_dtype(df):
+    """Ensure that struct function is only applied to correct dtype."""
+    with pytest.raises(TypeError):
```

Comment:
Odd, I'd expect a KeyError, I guess this is Arrow's exception we're propagating here. Also, if used with an integer, it should be an IndexError, they seem to have a super class (https://docs.python.org/3/library/exceptions.html#LookupError) should we catch those issues ourselves, and throw an appropriate exception, or would a TypeError make sense (since the keys are part of the type).

Reply:
In this specific test case I basically want to ensure that struct operations are only applied to expressions/columns holding a struct dtype. An appropriate exception message is given that the used method is not valid on non struct dtypes.

Concerning tests checking for non existing labels or index positions (e.g. `test_struct_get_invalid_field`), I agree that is more appropriate to use `LookupError` instead of `TypeError`. An appropriate exception message about valid lookup values is already given.

Reply:
@JovanVeljanoski strong opinions on this?

Reply:
No strong opinion, but my preference in line with the @mansenfranzen describes above. 

regarding the key, indices, key_indices and the like, also no strong preference, but I am more in line with @maartenbreddels's approach here. I'm in not bloating up the API too much, unless you think it is quite common to access and use these things. 

Finally, speaking of duplicate keys, when they exist, displaying the dataframe in Jupyter is a bit funny (dunno if that can be helped.. maybe you are already aware of it..).

<img width="1132" alt="image" src="https://user-images.githubusercontent.com/18574951/126633893-a1455dda-f937-4919-b782-8399abd31920.png">

Reply:
@maartenbreddels  @JovanVeljanoski 

Agreed, let's drop `key_indices` and explicitly state the assumption of correctly ordered field-index positions in docs.

Regarding the print output: it looks indeed a bit funny. Those backslashes are not present under my ubuntu 20.04 environment but I'll take a closer look at it. 

Anyhow, we have to agree on how print output should be formatted when duplicated fields are present. In my first attempt, I added the index position in parenthesis after the field name to distinguish duplicated fields. However, it looks odd to me right now. I assume there are different possibilities:

a) both field name and index position: `{'col1(0)': 1, 'col1(1)': "a", 'col3(2)': 4}`
b) fields only: `{'col1': 1, 'col1': "a", 'col3': 4}`
c) index only: `{0: 1, 1: "a", 2: 4}`

Currently I favor fields only because it is consistent even though duplicated fields occur. However, since we assume ordered field-index positions, one can still distinguish those values. What do you think?

Reply:
Hey, 

I am inclined to agree with you, regarding the fields only approach. I fear otherwise would be confusing, and we've had in the past people that had that of the type "column_name[x]" as much as it was surprising to me..

Reply:
@JovanVeljanoski What would be your favorite way to represent structs with duplicated field names?

Reply:
Fields only (option "b" from your previous message). If those are the names, we show them as they are.. 
In general I'd like to show/see the data as it is.

Reply:
@JovanVeljanoski Can you please check if you still have the strange formatting output for structs with duplicated fields?